### ISSUE TITLE
Could not rely on StateManagerChanged notification to build in memory projection of Reliable Collection

### ISSUE THREAD

**Issue by pmgfrederico** [Azure/service-fabric-issues#1367]
Dear sirs,

Is the Reliable Service notifications [documentation ](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-reliable-services-notifications)up to date with the regards to its expected behavior?

The use case I am after is about having a simple in memory repository that is projected from an existing reliable collection. I was expecting to upgrade the application and that with nodes coming down and up again Rebuild notifications would allow me to hydrate the repository seamlessly. The thing is that on the Local Cluster debugging with AutoUpgrade I more or less get the expected behavior but not on a remote production cluster.

Other findings are, Rebuild notifications don't ever have any ReliableStates sent with their NotifyStateManagerChangedEventArgs. Even when the RebuildNotificationAsyncCallback is called upon will the ReliableStates be shared via the notification's arguments. In fact it seems to trigger the NotifyStateManagerChangedAction.Add action and then run through the DictionaryChanged notifications as if it was the first seed of the data. This is very hard to follow, when the API suggest a given behavior that doesn't come true.

My question is, can I rely on Notifications for the aforementioned use case? Any ideas why the StateManager isn't behaving as expected?

Any help will be most appreciated.

Regards,
PF

**Comment by masnider** [comment]
You can rely and what you have described is the expected behavior. There's a basic sample [here](https://github.com/masnider/service-fabric-dotnet-data-aggregation-v2/blob/2319b0e90b337f1a4b0d0e5d54fd3802b64bc7cb/DataAggregation.NationalService/NationalService.cs) but I don't know that it is complete for the scenarios you're trying to cover. Can you share some snippets of your code showing how you're setting up the notifications and then where the inforamtion you're getting back isn't correct?

**Comment by pmgfrederico** [comment]
Hi.

Thanks for taking the time to investigate.

I've created a [gist](https://gist.github.com/pmgfrederico/a7ac64f66cbdb87a340fc2ce574e3f3f) with the implementation I'm using and it feels pretty much the same as the one you have shared on the aforementioned sample.

Just like in the sample I am subscribing the StateManagerChanged event on the StatefulService ctor and in development, as I previously mentioned, I kinda get the expected behavior. The one scenario that puzzles me is that when deploying to production (where the ReliabeDictionary I'm sourcing this projection from already exists and can't be wiped out) the StateManager won't run through the proper Dictionary Notifications after Rebuild - which I expect to happen when replicas are upgraded and need to reinstante the service's state.  

(The service is running on a single partition deployment with 3 instances.)

Can I share any other info that might help address the problem at hand?

Regards,
PF

**Comment by zuhairp** [comment]
You'll need to handle state manager rebuild as well, at which point you can add rebuild handlers for existing dictionaries. The documentation only briefly mentions it, and it's missing from the sample as well, so we'll need to address that documentation gap. 

Anyways, you'll add this to your StateManager_StateChanged method:

```C#
case NotifyStateManagerChangedAction.Rebuild:
    {
        var operation = e as NotifyStateManagerRebuildEventArgs;
        var reliableStatesEnumerator = operation.reliableStates.GetAsyncEnumerator();
        while (await reliableStatesEnumerator.MoveNextAsync(CancellationToken.None))
        {
            IReliableState reliableState = reliableStatesEnumerator.Current;
            // TODO: Add dictionary rebuild handler to reliableState as necessary
        }

        break;
    }
```
Please let us know if this helps. Thanks!

**Comment by pmgfrederico** [comment]
Hi @zuhairp 

Thanks for diving in.

I had similar code as the one you are suggesting (sadly the gist no longer represents that). That is the case I have described in the first comment, the ReliableStates won't ever return any state. Also await(ing) a void EventHandler...!?

Am I misreading something here?

**Comment by zuhairp** [comment]
Depending on the state of the system, you may get some or none of the ReliableStates in the rebuild notification, and the rest as individual SingleEntityChanged events. Just because it comes as an Add event does not mean that it is a newly created ReliableState, all the data you added to the state should still be there. Is that what you are seeing?

**Comment by pmgfrederico** [comment]
Sadly that's not what I am seeing.

When I upgrade the application on the Cluster the in-memory repository never gets re-hydrated despite the ReliableDictionary having its state consistent.

With regards to your previous statement I tried to locally trace StateManagerChanged events by bringing nodes up and down on the 5 node cluster and what I have witnessed is that even when a Primary Replica sees its StateManager Rebuild event triggered (e.g. restarting a node to force role change or for role change via powershell command) I can't enumerate across any of the ReliableStates that I expected to be made available via event args.

Almost giving up on this approach towards a persisted secondary index like collection.

Basically the solution works when I see a Rebuild followed by an Add action, but it doesn't when I see only and Rebuild action and nothing else. I am tracing this behavior using application insights on a time span neighborhood of deploying an application upgrade on a remote cluster.

**Comment by zuhairp** [comment]
Sorry for the delay in my response, I have been away. I would like to get to the bottom of this. Would you be able to open a support ticket through Azure portal for your production cluster? That way I can take a look through our SF traces to see what can be going on.

**Comment by juho-hanhimaki** [comment]
Would be great to get complete sample of using notifications (including rebuild and all possible other events that need to be handled). Documentation and samples are not great.
